Add benchmarks for SSLEngine implementations

normanmaurer opened the pull request:
Motivation:

As we provide our own SSLEngine implementation we should have benchmarks to compare it against JDK impl.

Modifications:

Add benchmarks for wrap / unwrap and handshake performance.

Result:

Benchmarks FTW.

normanmaurer commented:
@nmittler I took some inspiration from the work you did in conscrypt with your benchmarks but added a lot more then what you have. Maybe you want to port this over.

Scottmitch commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineThroughputBenchmark.java:
```diff
@@ -0,0 +1,86 @@
+
+    protected void doTearDown() throws Exception { }
+
+    @Setup(Level.Invocation)
```
have you seen the warning for `Level.Invocation` in the javadocs? Have you tried just putting this in the test ... if this consumes a relatively small proportion of the benchmark invocation resources (cpu/time) then maybe we will get more accurate results?

normanmaurer replied:
@Scottmitch yeah I saw this but I think we really not want to have this in the test itself.

Scottmitch replied:
Agreed that it is ideal from a logical perspective to have it completely isolated from the benchmark.  However I'm curious if we fall into the warnings that are listed in the javadocs for `Level.Invocation`.

As long as the overhead is consistent and "relatively small" compared to what we want to run in the benchmark then conceivably this may make the overhead of managing the counters/timers lower, reduce risk for coordinated omission, etc... WDYT?

normanmaurer replied:
@Scottmitch yeah good point. Let me see how it turns out when not using `Level.Invocation `

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,279 @@
+            sTOc.flip();
+
+            cTOsPos = cTOs.position();
+            sTOcPos = sTOc.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "sTOcPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,279 @@
+            }
+
+            if (!serverHandshakeFinished) {
+                int serverAppReadBufferPos = serverAppReadBuffer.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "serverAppReadBufferPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,279 @@
+        boolean serverHandshakeFinished = false;
+
+        do {
+            int cTOsPos = cTOs.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "cTOsPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,279 @@
+            sTOcPos = sTOc.position();
+
+            if (!clientHandshakeFinished) {
+                int clientAppReadBufferPos = clientAppReadBuffer.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "clientAppReadBufferPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,279 @@
+
+        do {
+            int cTOsPos = cTOs.position();
+            int sTOcPos = sTOc.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "sTOcPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,279 @@
+
+        private SslContext newServerContext() {
+            try {
+                File keyFile = new File(getClass().getResource("test_unencrypted.pem").getFile());
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Usage of GetResource in io.netty.microbench.handler.ssl.AbstractSslEngineBenchmark$SslEngineProvider.newServerContext() may be unsafe if class is extended [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AUI_INHERITANCE_UNSAFE_GETRESOURCE)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,279 @@
+
+        abstract ByteBuffer newBuffer(int size);
+
+        void freeBuffer(ByteBuffer buffer) { }
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1186)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,279 @@
+            cTOs.flip();
+            sTOc.flip();
+
+            cTOsPos = cTOs.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "cTOsPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

johnou commented:
@normanmaurer do you have some numbers or maybe a 3d bar graph to share? (everyone loves 3d bar graphs).

johnou commented:
Created a quick and dirty proto to plot the JMH results.. https://rawgit.com/johnou/ploty-proto/master/index.html

normanmaurer commented:
@Scottmitch PTAL again. Also changed the handshake benchmark to record the average time needed for a handshake. This may be a more useful benchmark

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/SslEngineEchoBenchmark.java:
```diff
@@ -0,0 +1,56 @@
+        ByteBuffer dst = unwrapDstBuffer;
+        dst.clear();
+
+        SSLEngineResult unwrapResult = serverEngine.unwrap(src, dst);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "unwrapResult". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins replied:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "unwrapResult". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,281 @@
+
+            cTOsPos = cTOs.position();
+            sTOcPos = sTOc.position();
+
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "sTOcPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,281 @@
+
+            if (!serverHandshakeFinished) {
+                int serverAppReadBufferPos = serverAppReadBuffer.position();
+                serverResult = serverEngine.unwrap(cTOs, serverAppReadBuffer);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "serverAppReadBufferPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,281 @@
+
+            if (!clientHandshakeFinished) {
+                int clientAppReadBufferPos = clientAppReadBuffer.position();
+                clientResult = clientEngine.unwrap(sTOc, clientAppReadBuffer);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "clientAppReadBufferPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,281 @@
+        do {
+            int cTOsPos = cTOs.position();
+            int sTOcPos = sTOc.position();
+
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "sTOcPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineThroughputBenchmark.java:
```diff
@@ -0,0 +1,81 @@
+        dst.clear();
+
+        SSLEngineResult wrapResult = clientEngine.wrap(src, dst);
+
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "wrapResult". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineThroughputBenchmark.java:
```diff
@@ -0,0 +1,81 @@
+        ByteBuffer dst = wrapDstBuffer;
+        dst.clear();
+
+        SSLEngineResult wrapResult = clientEngine.wrap(src, dst);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "wrapResult". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

Scottmitch commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,281 @@
+                assert empty.remaining() == clientResult.bytesConsumed();
+                assert cTOs.position() - cTOsPos == clientResult.bytesProduced();
+
+                if (isHandshakeFinished(clientResult)) {
```
to avoid another conditional consider just assigning unconditionally:

```java
clientHandshakeFinished = isHandshakeFinished(clientResult);
```

Scottmitch replied:
same comment below

normanmaurer replied:
+1

Scottmitch commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineThroughputBenchmark.java:
```diff
@@ -0,0 +1,81 @@
+        ByteBuffer src = wrapSrcBuffer;
+        src.position(0).limit(messageSize);
+
+        ByteBuffer dst = wrapDstBuffer;
```
are these temp variables necessary?

normanmaurer replied:
Not really just preferred this to fetch it multiple times from the fields. Should I remove ?

Scottmitch replied:
if they are not necessary then lets kill them ... less is more :)

Scottmitch commented on microbench/src/main/java/io/netty/microbench/handler/ssl/SslEngineHandshakeBenchmark.java:
```diff
@@ -0,0 +1,55 @@
+        initEngines();
+        boolean ok = doHandshake();
+        assert ok;
+        // We explicit not call destroyEngines() and let the GC handle this as its a NOOP for JDK but not for
```
`explicit not` -> `explicitly do not`

Scottmitch replied:
Is this shown to have a noticeable impact? Can we also test with `SslProvider#OPENSSL_REFCNT` (and cleanup after the iteration) to avoid this extra GC costs and see if finalizers have impacts? I'm curious if any differences exist and if they are noticeable with the short running micro benchmarks.

normanmaurer replied:
I did not test it. Let me add the destroy here and also add the reference counted one to the mix.

Scottmitch commented:
Here is what I see from a run on my local laptop:

```
Benchmark                          (bufferType)                               (cipher)  (messageSize)  (sslProvider)   Mode  Cnt        Score       Error  Units
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256             64            JDK  thrpt   20   137363.675 ±  1526.316  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256             64        OPENSSL  thrpt   20   309695.549 ±  3657.011  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256            128            JDK  thrpt   20   126470.268 ±   775.742  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256            128        OPENSSL  thrpt   20   271463.164 ±  1361.166  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256            512            JDK  thrpt   20    82317.191 ±   366.145  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256            512        OPENSSL  thrpt   20   163091.447 ±  3215.376  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           1024            JDK  thrpt   20    56161.760 ±   191.789  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           1024        OPENSSL  thrpt   20   121405.125 ±   586.676  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096            JDK  thrpt   20    19050.618 ±   132.286  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096        OPENSSL  thrpt   20    44067.334 ±   265.144  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             64            JDK  thrpt   20   263004.676 ±  1204.425  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             64        OPENSSL  thrpt   20   899465.746 ±  8956.711  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            128            JDK  thrpt   20   174997.924 ±   820.989  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            128        OPENSSL  thrpt   20   934805.942 ± 13000.018  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            512            JDK  thrpt   20    52961.139 ± 11129.199  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            512        OPENSSL  thrpt   20   738901.276 ± 13330.939  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           1024            JDK  thrpt   20    30847.091 ±   448.797  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           1024        OPENSSL  thrpt   20   576304.973 ± 22299.466  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096            JDK  thrpt   20     8114.233 ±    44.995  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096        OPENSSL  thrpt   20   251647.101 ±  3018.577  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256             64            JDK  thrpt   20   136138.740 ±  3549.283  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256             64        OPENSSL  thrpt   20   336008.345 ±  5581.356  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256            128            JDK  thrpt   20   123949.087 ±  2731.908  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256            128        OPENSSL  thrpt   20   277512.957 ±  6897.793  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256            512            JDK  thrpt   20    80528.069 ±   443.529  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256            512        OPENSSL  thrpt   20   174889.802 ±  2050.056  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           1024            JDK  thrpt   20    54305.089 ±   285.469  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           1024        OPENSSL  thrpt   20   125576.726 ±   609.402  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096            JDK  thrpt   20    18654.479 ±    89.622  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096        OPENSSL  thrpt   20    45750.151 ±   351.605  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             64            JDK  thrpt   20   253680.512 ±  3648.404  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             64        OPENSSL  thrpt   20  1187121.634 ± 10535.939  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            128            JDK  thrpt   20   169870.074 ±   928.175  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            128        OPENSSL  thrpt   20  1107897.661 ± 12700.059  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            512            JDK  thrpt   20    57019.398 ±   361.496  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256            512        OPENSSL  thrpt   20   871846.542 ±  5714.718  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           1024            JDK  thrpt   20    30208.858 ±   242.793  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           1024        OPENSSL  thrpt   20   699620.427 ±  6361.766  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096            JDK  thrpt   20     7841.974 ±    43.494  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096        OPENSSL  thrpt   20   306512.975 ±  2787.494  ops/s
```

normanmaurer commented:
@Scottmitch addressed your comments and these are the numbers (for 4096 only):

```
# Run complete. Total time: 00:08:30

Benchmark                          (bufferType)                               (cipher)  (messageSize)   (sslProvider)   Mode  Cnt       Score      Error  Units
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096             JDK  thrpt   20   16696.562 ±  307.690  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096         OPENSSL  thrpt   20   38452.147 ± 1279.212  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096  OPENSSL_REFCNT  thrpt   20   39068.671 ±  441.447  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096             JDK  thrpt   20    7173.097 ±   61.434  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096         OPENSSL  thrpt   20  222900.317 ± 3404.781  ops/s
SslEngineEchoBenchmark.wrapUnwrap          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096  OPENSSL_REFCNT  thrpt   20  222996.005 ± 2844.444  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096             JDK  thrpt   20   16451.711 ±  117.141  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096         OPENSSL  thrpt   20   38971.455 ± 2362.105  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256           4096  OPENSSL_REFCNT  thrpt   20   40129.699 ±  540.628  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096             JDK  thrpt   20    6969.103 ±   60.482  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096         OPENSSL  thrpt   20  272399.582 ± 1963.944  ops/s
SslEngineEchoBenchmark.wrapUnwrap        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256           4096  OPENSSL_REFCNT  thrpt   20  272260.835 ± 3234.789  ops/s
```

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,283 @@
+            cTOs.flip();
+            sTOc.flip();
+
+            cTOsPos = cTOs.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "cTOsPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,283 @@
+            }
+
+            if (!serverHandshakeFinished) {
+                int serverAppReadBufferPos = serverAppReadBuffer.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "serverAppReadBufferPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,283 @@
+        boolean serverHandshakeFinished = false;
+
+        do {
+            int cTOsPos = cTOs.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "cTOsPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

netkins commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,283 @@
+            sTOcPos = sTOc.position();
+
+            if (!clientHandshakeFinished) {
+                int clientAppReadBufferPos = clientAppReadBuffer.position();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "clientAppReadBufferPos". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)

normanmaurer commented:
@trustin you may be interested as well.

johnou commented:
@normanmaurer I would suggest using ```@OutputTimeUnit(TimeUnit.MICROSECONDS)``` for handshake otherwise the scores don't make much sense.

```
Benchmark                              (bufferType)                               (cipher)   (sslProvider)  Mode  Cnt  Score    Error  Units
SslEngineHandshakeBenchmark.handshake          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256             JDK  avgt   20  0.012 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256         OPENSSL  avgt   20  0.002 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake          HEAP  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256  OPENSSL_REFCNT  avgt   20  0.002 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             JDK  avgt   20  0.011 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256         OPENSSL  avgt   20  0.002 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake          HEAP  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256  OPENSSL_REFCNT  avgt   20  0.002 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256             JDK  avgt   20  0.012 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256         OPENSSL  avgt   20  0.002 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256  OPENSSL_REFCNT  avgt   20  0.002 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256             JDK  avgt   20  0.011 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256         OPENSSL  avgt   20  0.002 ±  0.001   s/op
SslEngineHandshakeBenchmark.handshake        DIRECT  TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256  OPENSSL_REFCNT  avgt   20  0.002 ±  0.001   s/op
```

normanmaurer commented:
@johnou good idea!

Scottmitch commented on microbench/src/main/java/io/netty/microbench/handler/ssl/AbstractSslEngineBenchmark.java:
```diff
@@ -0,0 +1,283 @@
+                assert empty.remaining() == serverResult.bytesConsumed();
+                assert sTOc.position() - sTOcPos == serverResult.bytesProduced();
+
+                if (isHandshakeFinished(serverResult)) {
```
to avoid another conditional consider just assigning unconditionally

same comment as https://github.com/netty/netty/pull/6360/files#r102627897

normanmaurer commented:
Cherry-picked into 4.1 (d73477c7bd1fe7665a382a90796efba87c95dd18) and 4.0 (95b4814e3a8bd338b5c663c3c80d76bfa00dc1da)
